manticore/core/smtlib/expression.py

```diff
@@ -400,555 +467,749 @@ def __ne__(self, other):
+        """Gets an element from an empty Array."""
+        index = self.cast_index(index)
+        if default is None:
+            default = self._default
+        if default is not None:
+            return default
+        return ArraySelect(self, index)
 
-        # If taint was not forced by a keyword argument, calculate default
-        kwargs.setdefault("taint", reduce(lambda x, y: x.union(y.taint), operands, frozenset()))
+    def select(self, index):
+        return self.get(index)
 
-        super().__init__(array.index_bits, array.index_max, array.value_bits, **kwargs)
+    def store(self, index, value):
+        index = self.cast_index(index)
+        value = local_simplify(self.cast_value(value))
+        return ArrayStore(array=self, index=index, value=value)
 
     @property
-    def operands(self):
-        return self._operands
+    def written(self):
+        return set()
+
+    def is_known(self, index):
+        return False
+
+    @property
+    def underlying_variable(self):
+        array = self
+        while not isinstance(array, ArrayVariable):
+            array = array.array
+        return array
+
+
+class ArrayOperation(Array, Operation, abstract=True):
+    """ It's an operation that results in an Array"""
+
+    pass
+
+
+def get_items(array):
+    if isinstance(array, ArrayStore):
+        yield from get_items_array_store(array)
+    elif isinstance(array, ArraySlice):
+        yield from get_items_array_slice(array)
+    elif isinstance(array, ArrayConstant):
+        yield from get_items_array_constant(array)
+    return
+
+
+def get_items_array_slice(array):
+    assert isinstance(array, ArraySlice)
+    for offset, value in get_items(array.array):
+        yield offset + array.offset, value
+
+
+def get_items_array_store(array):
+    assert isinstance(array, ArrayStore)
+    while isinstance(array, ArrayStore):
+        yield array.index, array.value
+        array = array.array
+    yield from get_items(array)
+
+
+def get_items_array_constant(array):
+    assert isinstance(array, ArrayConstant)
+    for index, value in enumerate(array.value):
+        yield index, value
+
+
+def get_items_array_variable(array):
+    assert isinstance(array, ArrayVariable)
+    raise GeneratorExit
 
 
 class ArrayStore(ArrayOperation):
-    def __init__(self, array: "Array", index: "BitVec", value: "BitVec", *args, **kwargs):
-        assert index.size == array.index_bits
-        assert value.size == array.value_bits
-        super().__init__(array, index, value, *args, **kwargs)
+    __xslots__: Tuple[str, ...] = (
+        "_written#v",
+        "_concrete_cache#v",
+        "_length#v",
+        "_default#v",
+    )
+
+    @property
+    def concrete_cache(self):
+        if self._concrete_cache is not None:
+            return self._concrete_cache
+        self._concrete_cache = {}
+        for index, value in get_items(self):
+            if not isinstance(index, Constant):
+                break
```

Comment:
Because.....
Once you find a symbolic write. Or once you found a write to a symbolic index this symbolic index could overwrite any(*) of the following Constants indexes that may come next.


* actually it could potentially overwrite some and not others. You could get smart here and try to partition the array somehow.